client/kafka/kafka.go

```diff
@@ -5,45 +5,73 @@ import (
 	"errors"
 	"fmt"
 
-	"github.com/Shopify/sarama"
 	"github.com/beatlabs/patron/correlation"
 	"github.com/beatlabs/patron/encoding"
-	"github.com/beatlabs/patron/trace"
+
+	"github.com/Shopify/sarama"
 	"github.com/opentracing/opentracing-go"
-	"github.com/opentracing/opentracing-go/ext"
 	"github.com/prometheus/client_golang/prometheus"
 )
 
 const (
-	producerComponent     = "kafka-async-producer"
-	messageCreationErrors = "creation-errors"
-	messageSendErrors     = "send-errors"
-	messageSent           = "sent"
+	asyncProducerComponent = "kafka-async-producer"
+	syncProducerComponent  = "kafka-sync-producer"
+	messageCreationErrors  = "creation-errors"
+	messageSendErrors      = "send-errors"
+	messageSent            = "sent"
 )
 
-var messageStatus *prometheus.CounterVec
+var (
+	asyncMessageStatus *prometheus.CounterVec
+	syncMessageStatus  *prometheus.CounterVec
+)
+
+// Producer interface for Kafka.
+type Producer interface {
+	Send(ctx context.Context, msg *Message) error
+	Close() error
+}
 
-func messageStatusCountInc(status, topic string) {
-	messageStatus.WithLabelValues(status, topic).Inc()
+type baseProducer struct {
+	cfg           *sarama.Config
+	prodClient    sarama.Client
+	tag           opentracing.Tag
+	enc           encoding.EncodeFunc
+	contentType   string
+	messageStatus *prometheus.CounterVec
+}
+
+var (
+	_ Producer = &AsyncProducer{}
```

Comment:
Is this for checking that the Async/sync producer implement the producer interface? if yes, we should move this to a unit test then.

Reply:
@mantzas why not both? These are called compile-time assertions, see https://medium.com/@matryer/golang-tip-compile-time-checks-to-ensure-your-type-satisfies-an-interface-c167afed3aae or https://github.com/golang/go/blob/go1.8rc2/src/io/multi.go#L72

Reply:
Since it seems that Go uses ti in their repository I am already conviced! ;)